File: javascript/bank-account/bank-account.js
export class BankAccount {
  #balance;
  #open;

  constructor() {
    this.#open = false;
    this.#balance = null;
  }

  open() {
    if (this.#open) {
      throw new ValueError();
    }

    this.#open = true;
    this.#balance = 0;
  }

  close() {
    if (!this.#open) {
      throw new ValueError();
    }

    this.#open = false;
  }

  deposit(amount) {
    if (!this.#open) {
      throw new ValueError();
    }

    if (amount < 0) {
      throw new ValueError();
    }

    this.#balance += amount;
  }

  withdraw(amount) {
    if (!this.#open) {
      throw new ValueError();
    }

    if (amount < 0 || amount > this.#balance) {
      throw new ValueError();
    }

    this.#balance -= amount;
  }

  get balance() {
    if (!this.#open) {
      throw new ValueError();
    }
    
    return this.#balance;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}

File: lua/bank-account/bank-account.lua
local BankAccount = {}
BankAccount.__index = BankAccount

function BankAccount.new()
  local self = setmetatable({}, BankAccount)
  self.b = 0
  self.closed = false

  return self
end

function BankAccount:balance()
  return self.b
end

function BankAccount:deposit(amount)
  if (amount <= 0) then
    error("Amount has to be positive")
  end
  if (self.closed) then
    error("Deposits not possible for closed accounts")
  end
  self.b = self.b + amount
end

function BankAccount:withdraw(amount)
  if (amount <= 0) then
    error("Amount has to be positive")
  end
  if (amount > self.b) then
    error("Cannot overdraw")
  end
  if (self.closed) then
    error("Withdrawls not possible for closed accounts")
  end
  self.b = self.b - amount
end

function BankAccount:close()
  self.closed = true
end

return BankAccount
